runtime/lua/vim/lsp/diagnostic.lua

```diff
@@ -1282,6 +1287,43 @@ function M.set_loclist(opts)
     vim.cmd [[lopen]]
   end
 end
+
+--- Disable diagnostics for the given buffer
+--- @param bufnr (optional, number): Buffer handle, defaults to current
+-- Note that when diagnostics are disabled for a buffer, the server will still
+-- send diagnostic information and the client will still process it. The
+-- diagnostics are simply not displayed to the user.
+function M.disable(bufnr)
```

Comment:
I think we should also optionally allow a client id to be passed in, and default to all clients attached to the buffer.